boring/src/ssl/callbacks.rs

```diff
@@ -19,31 +21,35 @@ use crate::ssl::{
 };
 use crate::x509::{X509StoreContext, X509StoreContextRef};
 
-pub extern "C" fn raw_verify<F>(preverify_ok: c_int, x509_ctx: *mut ffi::X509_STORE_CTX) -> c_int
+pub(super) unsafe extern "C" fn raw_verify<F>(
+    preverify_ok: c_int,
+    x509_ctx: *mut ffi::X509_STORE_CTX,
+) -> c_int
 where
     F: Fn(bool, &mut X509StoreContextRef) -> bool + 'static + Sync + Send,
 {
-    unsafe {
-        let ctx = X509StoreContextRef::from_ptr_mut(x509_ctx);
-        let ssl_idx = X509StoreContext::ssl_idx().expect("BUG: store context ssl index missing");
-        let verify_idx = SslContext::cached_ex_index::<F>();
-
-        // raw pointer shenanigans to break the borrow of ctx
-        // the callback can't mess with its own ex_data slot so this is safe
-        let verify = ctx
-            .ex_data(ssl_idx)
-            .expect("BUG: store context missing ssl")
-            .ssl_context()
-            .ex_data(verify_idx)
-            .expect("BUG: verify callback missing") as *const F;
-
-        (*verify)(preverify_ok != 0, ctx) as c_int
-    }
+    // SAFETY: boring provides valid inputs.
+    let ctx = unsafe { X509StoreContextRef::from_ptr_mut(x509_ctx) };
+
+    let ssl_idx = X509StoreContext::ssl_idx().expect("BUG: store context ssl index missing");
+    let verify_idx = SslContext::cached_ex_index::<F>();
+
+    let verify = ctx
+        .ex_data(ssl_idx)
+        .expect("BUG: store context missing ssl")
+        .ssl_context()
```

Comment:
Should we have `to_owned()` call here?

Reply:
I explained just below in a `SAFETY` note why there is no need to own the context.